### PR TITLE
Checkout: add ID Wallet (OVO) payment method

### PR DESCRIPTION
**Pull request by michaeldcain**
This adds Indonesian Wallet (OVO) support to the new Checkout flow. The payment method is currently disabled for both Checkouts, while improvements are made to the refund flow.

Screen Shot 2020-08-28 at 10 42 57 AM

**To Test:**

- `id_wallet` is already available to a12s, but will need to be force if testing a non-a11n account
- set your currency to IDR (Indonesia)
- visit Checkout
- verify that OVO is shown as a payment method
- verify that name, NIK, and phone inputs (with Indonesia pre-selected) are shown
- verify that all 3 fields are required
- submit a payment with any name, 16-digit NIK, and 10-digit phone number
- verify that you're redirected to dLocal's sandbox payment gateway

Fixes https://github.com/Automattic/wp-calypso/issues/44832

### CODE REVIEW

File: client/my-sites/checkout/composite-checkout/types/backend/payment-method.ts [hunk 1]
```diff
@@ -196,6 +203,8 @@ export function translateCheckoutPaymentMethodToWpcomPaymentMethod(
 			return { name: 'WPCOM_Billing_Ebanx_Redirect_Brazil_Tef' };
 		case 'netbanking':
 			return { name: 'WPCOM_Billing_Dlocal_Redirect_India_Netbanking' };
+		case 'id-wallet':
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/types/backend/payment-method.ts:206]: As I'd like to get rid of these functions in favor of their already-existing counterparts in calypso, I think we should use the same keys as the existing ones. In #40049 you'll see that the key for this is `id_wallet`, with an underscore, so we should probably change all of these to use that key instead.

File: client/my-sites/checkout/composite-checkout/payment-methods/id-wallet.js [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+	} );
+
+	return { ...store, actions, selectors };
+}
+
+export function createIdWalletMethod( { store, stripe, stripeConfiguration } ) {
+	return {
+		id: 'id-wallet',
+		label: <IdWalletLabel />,
+		activeContent: <IdWalletFields stripe={ stripe } stripeConfiguration={ stripeConfiguration } />,
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/payment-methods/id-wallet.js:131]: As you realized while we were pairing on this, if you look below you'll see that `IdWalletFields` does not use `stripe` or `stripeConfiguration`, so we can remove those props from here and from this function.

File: client/my-sites/checkout/composite-checkout/payment-methods/id-wallet.js [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+		submitButton: (
+			<IdWalletPayButton
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/payment-methods/id-wallet.js:133]: Same for `IdWalletPayButton`; it does not need the `stripe` or `stripeConfiguration` data.

File: client/my-sites/checkout/composite-checkout/payment-methods/id-wallet.js [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+				label={ __( 'Your name' ) }
+				value={ customerName?.value ?? '' }
+				onChange={ changeCustomerName }
+				isError={ customerName?.isTouched && customerName?.value.length === 0 }
+				errorMessage={ __( 'This field is required' ) }
+				disabled={ isDisabled }
+			/>
+			<div className="id-wallet__contact-fields">
+				<CountrySpecificPaymentFieldsUI
+					countryCode={ 'ID' } // If this payment method is available and the country is not India, we have other problems
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/payment-methods/id-wallet.js:180]: This string is correct, but the comment is not. Might as well remove the comment, I guess?

**Comment by nbloomf** [comment]
Here are the results of my testing. I ran into one bug, one possible bug, and a couple of non-functional enhancements.

- I set my currency to INR and entered checkout using the new site flow.
- The `WPCOM_Billing_Dlocal_Redirect_Indonesia_Wallet` payment method is available in responses from the shopping cart endpoint, as expected.
- OVO is available in the list of payment methods. I may be imagining it, but it looks like the label has _just barely_ more whitespace at the bottom than at the top. That should be an easy fix.

image

- Selecting the OVO method shows the extra fields. The phone number field is shorter than the others - whether that's a problem is a decision needing input from a designer.

image

- Attempting to pay without filling in any fields shows the field errors as expected, and the errors are cleared when I start editing. For good measure I filled in the fields with junk data and went back to edit the other checkout steps; my OVO data persisted, as expected.
- I tried a few different invalid NIKs. Validation here seems to be working correctly, but there are a couple of UX improvements we could make.
    - If the length is wrong we show a pretty generic error message. Our other error messages indicate what the correct format should be; it might be good to to that here too. It's surprisingly difficult to find english language information on what the format of this should be ("customs identification number" or "Nomor Induk Kependudukan") but it appears to be 16 decimal digits.
    - Without looking at the code, I suspect we're stripping out non-digit characters before validating, but possibly not before sending to dlocal. the code `1212121212121212a` passes validation (it has 16 digits) but causes an error notification but does not put an error on the field. **I think this is a bug**.

image

- **If the transfer expires**: I am returned to checkout, as expected, and everything works. There is no notification that the payment failed (**maybe a bug**), and my payment info is not persisted.
- **If the transfer succeeds**: I return to the post-checkout flow as expected. The site provisions and the receipt generates correctly.

**Comment by fditrapani** [comment]
It looks like the OVO logo is adding a couple extra pixels to the option. I can take a look tonight to clean it up.

**Comment by fditrapani** [comment]
Just pushed an update for the logo. Here's a before:

image

and after:
image

**Comment by sirbrillig** [comment]
Excellent testing as always, @nbloomf 

> Without looking at the code, I suspect we're stripping out non-digit characters before validating, but possibly not before sending to dlocal. the code 1212121212121212a passes validation (it has 16 digits) but causes an error notification but does not put an error on the field. I think this is a bug.

So it sounds like we have three options: A) make sure we don't strip non-digits during validation (so your example would have failed before sending), or B) strip non-digits when sending also, and/or C) properly show the error notification from the transaction endpoint on the failing field. Option C is a good idea but challenging as we don't currently have any system in place to absorb transaction errors and turn them into field errors in new checkout, so I think that it should wait until we have that set up. While it feels more accurate to do option A, I think that option B is going to be a smoother UX. I checked `validators.validIndonesiaNik()` and [it does indeed strip out the non-digits](https://github.com/Automattic/wp-calypso/blob/eb6d12ec1e6a5016481568c3559e854bf4bdfa49/client/lib/checkout/validation.js#L295). While it's hard to search for this, apart from validation I don't see that behavior anywhere in old checkout so I think we should just copy/paste the regexp from that validation function (`value.replace( /[^0-9]/g, '' )`) into the new payment method during or just after we create `massagedFields`.

> If the transfer expires: I am returned to checkout, as expected, and everything works. There is no notification that the payment failed (maybe a bug)

This seemed surprising, but when I thought about it, it seems expected for all our redirect payment methods (eg: PayPal). We provide a return URL to each of the redirects so they can get back to checkout if they fail, but IIRC there's no mechanism in that URL to display an error message. I tested this with old checkout with PayPal, and it appears that old checkout has this issue as well, so I'm going to say we don't need to fix this here.

**Comment by sirbrillig** [comment]
> Just pushed an update for the logo. Here's a before:

@fditrapani I don't think you actually pushed it. I don't see any commits in here. 🤔

**Comment by fditrapani** [comment]
Sorry about that @sirbrillig. I got a weird terminal error when I pushed and didn't notice it. I pushed the changes but it looks like something else came in. Not sure if that's meant to be there?

**Comment by sirbrillig** [comment]
No problem, it's just a merge commit because of my masking changes that made it up before yours. I'll rebase it out.

**Comment by fditrapani** [comment]
Thanks Payton!

**Comment by sirbrillig** [comment]
Note: I discovered a bug with the NIK input field and masking as I was adding the masking: https://github.com/Automattic/wp-calypso/pull/45311

**Pull request merged**